19.0 tutorial module

abdrahmanrashed opened the pull request.

abdrahmanrashed committed:
[IMP] -added some estate property types that are available with estate module installation.

abdrahmanrashed committed:
[IMP] -added some demo estate model data for estate module.

abdrahmanrashed committed:
[IMP] -Added some demo estate offer model data for estate module.

abdrahmanrashed committed:
[IMP] -Set demo estate offers dates to be relative to module creation time.

abdrahmanrashed committed:
[IMP] -Set demo estate properties to be residential type.

abdrahmanrashed committed:
[IMP] -Set an offer to be accepted by default, and the rest to be refused.

abdrahmanrashed committed:
[IMP] -Created demo estate property with offers inside.

vava-odoo commented on estate/demo/estate_property.xml:
```diff
@@ -0,0 +1,61 @@
+        <field name="state">new</field>
+        <field name="description">A nice and big villa</field>
+        <field name="postcode">12345</field>
+        <field name="date_availability">2020-02-02</field>
```
This should preferably be relative to the date of install (see codebase for examples)

vava-odoo commented on estate/demo/estate_property.xml:
```diff
@@ -0,0 +1,61 @@
+<odoo>
+    <record id="estate_property_1" model="estate.property">
+        <field name="name">Big Villa</field>
+        <field name="state">new</field>
```
As it is the default value, you're not forced to provide it
```suggestion
```

vava-odoo commented on estate/demo/estate_property.xml:
```diff
@@ -0,0 +1,61 @@
+        <field name="postcode">54321</field>
+        <field name="date_availability">1970-01-01</field>
+        <field name="expected_price">100000</field>
+        <field name="selling_price">120000</field>
```
Why a selling_price if the state is cancelled?

vava-odoo commented on estate/demo/estate_property_offer.xml:
```diff
@@ -0,0 +1,29 @@
+    <function model="estate.property.offer" name="action_offer_refuse" eval="[ref('estate_property_offer_1')]"/>
+    <function model="estate.property.offer" name="action_offer_refuse" eval="[ref('estate_property_offer_2')]"/>
+    <function model="estate.property.offer" name="action_offer_accept" eval="[ref('estate_property_offer_3')]"/>
+</odoo>
```
Missing EOF line-break

vava-odoo reviewed (commented):
Pretty good :+1:

abdrahmanrashed committed:
[ADD] -Added check to raise an error if an offer is created for a sold/cancelled property.
-Added check to raise an error if a property is sold without any offer accepted.
-Created tests cases to verify total area calculation, verify action_property_sold, and verify offer creation for sold property.
-Added missing new lines in some files.

abdrahmanrashed committed:
[IMP] -Added tests to check garden values are set/reset correctly onchange.

vava-odoo commented on estate/tests/test_estate_property.py:
```diff
@@ -0,0 +1,105 @@
+
+        cls.properties = cls.env['estate.property'].create([{'name': 'test_house'}])
+        cls.partner = cls.env['res.partner'].create([{
+            'id': 'test_partner',
```
id?! don't touch this :smile: 
```suggestion
```

vava-odoo commented on estate/tests/test_estate_property.py:
```diff
@@ -0,0 +1,105 @@
+            'state_id': cls.env.ref('base.state_us_39').id,
+            'phone': '+1 555-555-5555',
+            'email': 'test@testing.example.com',
+            'tz': 'Europe/Brussels',
```
Do you need all this? I bet a name is sufficient for your case
```suggestion
```

vava-odoo commented on estate/tests/test_estate_property.py:
```diff
@@ -0,0 +1,105 @@
+        # add env on cls and many other things
+        super().setUpClass()
+
+        cls.properties = cls.env['estate.property'].create([{'name': 'test_house'}])
```
Plural is misleading
```suggestion
        cls.property = cls.env['estate.property'].create([{'name': 'test_house'}])
```

vava-odoo commented on estate/tests/test_estate_property.py:
```diff
@@ -0,0 +1,105 @@
+
+        self.assertRecordValues(self.properties, [
+           {'total_area': 35},
+        ])
```
assertEqual if only one value to check, no?

vava-odoo commented on estate/tests/test_estate_property.py:
```diff
@@ -0,0 +1,105 @@
+            'validity': 14,
+        })
+        self.properties.offer_ids.action_offer_accept()
+        self.properties.action_property_sold()
```
Not related to tests, but I have the feeling that action_offer_accept should mark the property as sold, so why do you need the call to action_property_sold?

abdrahmanrashed replied:
In the estate tutorial accepting the offer and selling the property are two different states. 
After accepting an offer you can still cancel the property, not sell it.

vava-odoo commented on estate/tests/test_estate_property.py:
```diff
@@ -0,0 +1,105 @@
+    def test_action_sell_with_accepted_offer(self):
+        """Test that everything behaves like it should when selling a valid property."""
+
+        self.properties.offer_ids.create({
```
Weird way to create an offer. Either use self.env['property.offer'].create(...), either create it while editing the field offer_ids with Command

vava-odoo reviewed (commented):
Looks quite ok :+1:

abdrahmanrashed committed:
[IMP] -refactoring based of PR feedback.
